File: src/data/courses/dsa/searching-algorithms/two-sorted-arrays/kth-element-two-sorted-arrays/code/python/kth_element_two_sorted_arrays.py
def kth_element(nums1, nums2, k):
    """
    Find kth element of two sorted arrays
    Time: O(log(min(m, n))), Space: O(1)
    """
    m, n = len(nums1), len(nums2)
    
    if m > n:
        return kth_element(nums2, nums1, k)
    
    low = max(0, k - n)
    high = min(k, m)
    
    while low <= high:
        cut1 = (low + high) // 2
        cut2 = k - cut1
        
        left1 = float('-inf') if cut1 == 0 else nums1[cut1 - 1]
        right1 = float('inf') if cut1 == m else nums1[cut1]
        left2 = float('-inf') if cut2 == 0 else nums2[cut2 - 1]
        right2 = float('inf') if cut2 == n else nums2[cut2]
        
        if left1 <= right2 and left2 <= right1:
            return max(left1, left2)
        elif left1 > right2:
            high = cut1 - 1
        else:
            low = cut1 + 1
    
    return -1

if __name__ == "__main__":
    nums1 = [2, 3, 6, 7, 9]
    nums2 = [1, 4, 8, 10]
    k = 5
    
    result = kth_element(nums1, nums2, k)
    print(f"Kth element: {result}")  # Output: 6

File: src/data/courses/dsa/searching-algorithms/two-sorted-arrays/kth-element-two-sorted-arrays/code/java/KthElementTwoSortedArrays.java
// Time: O(log(min(m, n)))
// Space: O(1)

public class KthElementTwoSortedArrays {
    /**
     * Find kth element of two sorted arrays
     */
    public static int kthElement(int[] nums1, int[] nums2, int k) {
        int m = nums1.length, n = nums2.length;
        
        if (m > n) {
            return kthElement(nums2, nums1, k);
        }
        
        int low = Math.max(0, k - n);
        int high = Math.min(k, m);
        
        while (low <= high) {
            int cut1 = (low + high) / 2;
            int cut2 = k - cut1;
            
            int left1 = (cut1 == 0) ? Integer.MIN_VALUE : nums1[cut1 - 1];
            int right1 = (cut1 == m) ? Integer.MAX_VALUE : nums1[cut1];
            int left2 = (cut2 == 0) ? Integer.MIN_VALUE : nums2[cut2 - 1];
            int right2 = (cut2 == n) ? Integer.MAX_VALUE : nums2[cut2];
            
            if (left1 <= right2 && left2 <= right1) {
                return Math.max(left1, left2);
            } else if (left1 > right2) {
                high = cut1 - 1;
            } else {
                low = cut1 + 1;
            }
        }
        
        return -1;
    }
    
    public static void main(String[] args) {
        int[] nums1 = {2, 3, 6, 7, 9};
        int[] nums2 = {1, 4, 8, 10};
        int k = 5;
        
        int result = kthElement(nums1, nums2, k);
        System.out.println("Kth element: " + result); // Output: 6
    }
}
